tensorflow/security/fuzzing/cc/ops/string_ops_fuzz.cc

```diff
@@ -0,0 +1,118 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include <gtest/gtest.h>
+
+#include <string>
+
+#include "fuzztest/fuzztest.h"
+#include "tensorflow/cc/ops/standard_ops.h"
+#include "tensorflow/security/fuzzing/cc/fuzz_session.h"
+
+namespace tensorflow {
+namespace fuzzing {
+
+class FuzzStringOpsStringSplit : public FuzzSession<std::string, std::string> {
+  void BuildGraph(const Scope& scope) override {
+    auto op_node =
+        tensorflow::ops::Placeholder(scope.WithOpName("input"), DT_STRING);
+    auto op_node2 =
+        tensorflow::ops::Placeholder(scope.WithOpName("delimiter"), DT_STRING);
+
+    tensorflow::ops::StringSplit(scope.WithOpName("output"), op_node, op_node2);
+  }
+
+  void FuzzImpl(const std::string& input_string,
+                const std::string& separator_string) final {
+    Tensor input_tensor(tensorflow::DT_STRING, {2});
+
+    auto svec = input_tensor.flat<tstring>();
+    svec(0) = input_string.c_str();
+    svec(1) = input_string.c_str();
+
+    Tensor separator_tensor(tensorflow::DT_STRING, TensorShape({}));
+    separator_tensor.scalar<tensorflow::tstring>()() = separator_string;
+
+    Status s = RunInputsWithStatus(
+        {{"input", input_tensor}, {"delimiter", separator_tensor}});
+    if (!s.ok()) {
+      LOG(ERROR) << "Execution failed: " << s.error_message();
+    }
+  }
+};
+FUZZ_TEST_F(FuzzStringOpsStringSplit, Fuzz)
+    .WithDomains(fuzztest::OneOf(fuzztest::InRegexp("[-.0-9]+"),
```

Comment:
Is the idea here to give a higher chance for strings of the first shape to be tested?

Reply:
The goal is just to provide fuzzed input to each of the two arguments to the ops -- I started with regex perhaps mostly because I was experimenting. Am happy to leave it as is, but we could also convert it to standard strings, i.e. not limit to the space of the regex -- or maybe just submit a new fuzzer with more types.

The main idea is to give the first shape the right type, such that the code is executed -- which it currently isn't. As such, currently there is zero chance for the first shape to be tested since they always fail this line:
```OP_REQUIRES(ctx, TensorShapeUtils::IsVector(input_tensor->shape()),```

See the code coverage report here: https://storage.googleapis.com/oss-fuzz-coverage/tensorflow/reports/20230225/linux/proc/self/cwd/tensorflow/core/kernels/string_split_op.cc.html#L177

I've verified locally the new fuzzers overcome the `OP_REQUIRES(ctx, TensorShapeUtils::IsVector(input_tensor->shape())` condition